cpp/src/io/parquet/bloom_filter_reader.cu

```diff
@@ -0,0 +1,680 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "arrow_filter_policy.cuh"
+#include "compact_protocol_reader.hpp"
+#include "io/parquet/parquet.hpp"
+#include "reader_impl_helpers.hpp"
+
+#include <cudf/ast/detail/expression_transformer.hpp>
+#include <cudf/ast/detail/operators.hpp>
+#include <cudf/ast/expressions.hpp>
+#include <cudf/detail/cuco_helpers.hpp>
+#include <cudf/detail/transform.hpp>
+#include <cudf/detail/utilities/logger.hpp>
+#include <cudf/hashing/detail/xxhash_64.cuh>
+#include <cudf/utilities/span.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_checks.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_buffer.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <cuco/bloom_filter_ref.cuh>
+#include <thrust/iterator/counting_iterator.h>
+
+#include <future>
+#include <numeric>
+#include <optional>
+
+namespace cudf::io::parquet::detail {
+namespace {
+
+/**
+ * @brief Converts bloom filter membership results (for each column chunk) to a device column.
+ *
+ */
+struct bloom_filter_caster {
+  cudf::device_span<cudf::device_span<cuda::std::byte> const> bloom_filter_spans;
+  host_span<Type const> parquet_types;
+  size_t total_row_groups;
+  size_t num_equality_columns;
+
+  enum class is_int96_timestamp : bool { YES, NO };
+
+  template <typename T, is_int96_timestamp IS_INT96_TIMESTAMP = is_int96_timestamp::NO>
+  std::unique_ptr<cudf::column> query_bloom_filter(cudf::size_type equality_col_idx,
+                                                   cudf::data_type dtype,
+                                                   ast::literal const* const literal,
+                                                   rmm::cuda_stream_view stream,
+                                                   rmm::device_async_resource_ref mr) const
+  {
+    using key_type    = T;
+    using policy_type = cuco::arrow_filter_policy<key_type, cudf::hashing::detail::XXHash_64>;
+    using word_type   = typename policy_type::word_type;
+
+    // Check if the literal has the same type as the predicate column
+    CUDF_EXPECTS(
+      cudf::have_same_types(cudf::column_view{dtype, 0, {}, {}, 0, 0, {}},
+                            cudf::column_view{literal->get_data_type(), 0, {}, {}, 0, 0, {}}),
```

Comment:
We can't compare `dtype` and `literal->get_data_type()` directly here?

Reply:
Bradley recommended that it is good practice `cudf::have_same_types` instead of directly comparing them.

Reply:
Using the helper function guarantees that nested types and other edge cases are handled properly.

However, this is currently throwing away the useful information. You have to compare an _actual_ column view and/or scalar, not a “fake” column view constructed from its dtype. Dtype equality is not sufficient for nested types. You need to be able to walk the nested child columns to compare dtypes properly, which is why the function takes columns/scalars instead of raw dtypes.

Reply:
Hmm. A couple of things here, we don't call this function for `cudf::is_compound<T>` types except `cudf::string_view` so we should be safe in regards to children (nesting), even with a direct compare. Also, we don't really have any columns materialized at this point and only have output column dtypes.

Reply:
You have a real scalar from the literal though, right? Maybe we can “fake” the column side but use the real scalar.

Reply:
@bdice and I chatted a bit. I am reverting to direct dtype compare until we find a way to get a usable `scalar` from the `literal`